Drop indexer changelog tables on mview unsubscribe

fredden opened the pull request:
### Description (*)
This pull request will remove the unused changelog tables when changing an index from 'update on schedule' to 'update on save'.

From what I can tell, this was the case before 6ae3c48d4df28a20590f7fe5d5deca472fd06084 was merged. I have tried to avoid any regressions here by marking the index as invalid when dropping the table, to catch any entries which might have otherwise been missed. It seems that 6ae3c48d4df28a20590f7fe5d5deca472fd06084 only protects missed changelog entries if the index is switched back to 'update on schedule'. The change introduced here solves that problem by always performing a full index (via normal minutely cron).

### Fixed Issues
- Fixes https://github.com/magento/ece-tools/issues/729
-  resolves magento/magento2#29789: Drop indexer changelog tables on mview unsubscribe

### Manual testing scenarios (*)
1. Set indexes to 'update on schedule'
1. Make a change which will trigger an entry in the changelog table (eg, update product name in admin)
1. Wait for normal minutely cron to run index
1. Observe data in `version_id` column in `mview_state` table.
1. Change indexes to 'update on save'
1. Observe data in `version_id` column in `mview_state` table.
1. Note that all changed indexes are now 'invalid' and will be fully indexed via normal minutely cron.

### Questions or comments
I am not confident that I have fully covered all changes here in unit tests. Please may I have some advice as to how I might need to adjust the unit tests to ensure no regressions.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#29789: Drop indexer changelog tables on mview unsubscribe

fredden committed:
Drop changelog table on indexer unsubscribe

fredden committed:
Invalidate index on mview unsubscribe

This means that the issue solved in MAGETWO-51078 (6ae3c48d4df28a20590f7fe5d5deca472fd06084) is not reintroduced by this changeset.

fredden committed:
Use provided interface for changing scheduled flag

fredden committed:
Update test to match new behaviour

kandy commented on app/code/Magento/Indexer/Setup/RecurringData.php:
```diff
@@ -49,7 +49,8 @@ public function install(ModuleDataSetupInterface $setup, ModuleContextInterface
             $indexer = $this->indexerFactory->create()->load($indexerId);
             if ($indexer->isScheduled()) {
-                $indexer->getView()->unsubscribe()->subscribe();
+                $indexer->setScheduled(false);
```
Please test that this not cause change log cleanup

fredden replied:
Yes, this will cause the changelog tables to be removed as it will call `$indexer->getView()->unsubscribe()` internally. Please can you help me understand the purpose of this recurring setup task if that behaviour is not desired?

fredden replied:
Further to this, I have discovered (and documented in a code comment) the purpose of this line. I've preserved the behaviour here - no changelog tables will be cleaned up when running this recurring setup script.

fredden committed:
Change documentation inheritance tag for linter

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into indexer-unsubscribe-drop

fredden committed:
Merge branch '2.4-develop' into indexer-unsubscribe-drop

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into indexer-unsubscribe-drop

fredden committed:
Reset mview status on unsubscribe

This fixes a problem where mview gets "stuck" if interrupted, for example
during deployment.

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into indexer-unsubscribe-drop

ihor-sviziev committed:
Merge branch '2.4-develop' into indexer-unsubscribe-drop

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
Hi @fredden,
Could you review and fix failing tests?

ihor-sviziev commented:
@magento create issue

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into indexer-unsubscribe-drop

fredden committed:
Fix tests

fredden commented:
@magento run all tests

sidolov commented:
Risk: high since the changes may affect the caching mechanism.

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8165](https://jira.corp.magento.com/browse/ENGCOM-8165) has been created to process this Pull Request

engcom-Bravo commented:
:heavy_check_mark: QA Passed

### Manual testing scenario
1. Go to **System - Index Management** 
2. **Select All** indexers from _Mass Actions_and select **Update by Schedule** from _Actions_ 
3. Click **Submit**
the indexers Status is updated, run `php bin/magento indexer:status`
![status1](https://user-images.githubusercontent.com/51680745/94560398-465a2080-026b-11eb-826f-c09ae1b209dd.png)
4. Go to **Catalog - Products** and **Edit** an existing _Product Name_. **Save** product
5. Run `php bin/magento indexer:status`
![status2](https://user-images.githubusercontent.com/51680745/94560638-9cc75f00-026b-11eb-8359-d4dddc97df45.png)
6. Wait for normal minutely cron to run index or run `php  bin/magento cron:run`
7. Check the indexer status `php bin/magento indexer:status`
![status3](https://user-images.githubusercontent.com/51680745/94561205-60483300-026c-11eb-93a9-bdfaf8769eb3.png)
8. Go to **System - Index Management** 
9. **Select All** indexers from _Mass Actions_and select **Update by Schedule** from _Actions_. Click **Submit**
10. Check the indexers Status, run `php bin/magento indexer:status`

**BEFORE** switching to the PR the indexer _Status_ was **Ready**
![rei_not_required](https://user-images.githubusercontent.com/51680745/94561433-a9988280-026c-11eb-9009-e635629bbc4e.png)

**AFTER** switching to the PR the indexer _Status_ was **Reindex required**
![rei_required](https://user-images.githubusercontent.com/51680745/94561503-c7fe7e00-026c-11eb-829b-73371ea662e1.png)

gabrieldagama commented:
@magento run Static Tests

ihor-sviziev commented:
@gabrieldagama is there any reasons why this PR isn't getting merged?

gabrieldagama commented:
@ihor-sviziev Yes, we are having an internal problem when trying to deliver this PR, there is an internal ticket to solve that, once that is resolved we can proceed with the merge.

fredden commented:
@gabrieldagama do I need to resolve the conflicts that are now showing on this pull request? Please can you provide an update on when this will be merged.

ihor-sviziev commented:
@fredden, yeah, could you resolve the conflict?

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into indexer-unsubscribe-drop

fredden commented:
@magento run all tests

ihor-sviziev commented:
@magento run Database Compare, Functional Tests B2B, Functional Tests CE, Functional Tests EE

Commit:
Merge branch '2.4-develop' into indexer-unsubscribe-drop

gabrieldagama commented:
@magento run all tests

gabrieldagama commented:
Hi @fredden, the problem mentioned before was fixed, but as this PR is low priority we will be processing and delivering high priority pull request first. This will be processed as soon as possible considering the priorities.

ihor-sviziev commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

ihor-sviziev commented:
@magento run Functional Tests B2B

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into indexer-unsubscribe-drop

fredden commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into indexer-unsubscribe-drop

fredden committed:
Avoid unnecessary invalidation on setup:upgrade

fredden commented:
@magento run all tests

fredden committed:
Define const visibility

fredden commented:
@magento run all tests

fredden committed:
Fix unit test

fredden committed:
Add PHP Code Mess ignore line, per other tests

fredden commented:
@magento run all tests

fredden commented:
@magento run Functional Tests B2B, Functional Tests EE

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8165](https://jira.corp.magento.com/browse/ENGCOM-8165) has been created to process this Pull Request

kandy commented on lib/internal/Magento/Framework/Mview/View.php:
```diff
@@ -219,21 +219,32 @@ public function subscribe()
+     * @inheritDoc
      */
-    public function unsubscribe()
+    public function unsubscribe(bool $dropTable = true): ViewInterface
```
Make sense to have `false` as default value to preserve BC

fredden replied:
Hi @kandy. I'm assuming that by 'BC' you mean "backwards compatibility."

If the parameter was required (ie, had no default value), it would break backwards compatibility. As the parameter has a default value, anything that calls this method without any arguments will continue to work. I therefore see this as being compatible with previous code, and not breaking backwards compatibility at a code level.

I think you're trying to ask for this method to behave the same as it does currently when called without any arguments. This parameter only exists so that it can be used in the recurring setup script where the intention is to recreate all triggers. If it weren't for that requirement, the parameter would not exist at all.

Can you provide a use-case where it would be useful to keep the changelog table after calling `mview->unsubscribe()`?

fredden committed:
Merge branch '2.4-develop' into indexer-unsubscribe-drop

fredden commented:
@magento run all tests

engcom-Lima commented:
@magento run Functional Tests B2B , Functional Tests EE , Semantic Version Checker

engcom-Lima commented:
This PR is already QA passed  referring [QA pass comment above](https://github.com/magento/magento2/pull/25859#issuecomment-700686833). Builds are failed hence, moving this PR to extended testing.

engcom-Lima commented:
@magento create issue

engcom-Charlie commented:
@engcom-Lima as the above mentioned QA [testing](https://github.com/magento/magento2/pull/25859#issuecomment-700686833) has been done long back, requesting you to go through the QA process again. Hence moving it back to Ready For Testing.

Thank you!

engcom-Lima commented:
:heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento `2.4-develop`

**Manual testing scenario:**

1. Go to System - Index Management
2. Select All indexers from _Mass Actions_ and select **`Update by Schedule`** from Actions
3. Click Submit
4. The indexers Status is updated, run `php bin/magento indexer:status`
5. Go to Catalog - Products and Edit an existing Product Name. Save product
6. Run `php bin/magento indexer:status`
7. Wait for normal minutely cron to run index or run `php  bin/magento cron:run`
8. Check the indexer status `php bin/magento indexer:status`
9. Go to System - Index Management
10. Select All indexers from _Mass Actions_ and select **`Update on save`** from Actions. Click Submit
11. Check the indexers Status, run `php bin/magento indexer:status`

**Before:** :heavy_multiplication_x:  Switching to the PR the indexer Status was **Ready**
<img width="1476" alt="Screenshot 2023-01-17 at 6 16 25 PM" src="https://user-images.githubusercontent.com/60198022/213406876-7ad9b30e-561d-40c7-9d35-12f6e05c4b7e.png">


**After:** :heavy_check_mark:  Switching to the PR the indexer Status was **Reindex required**

<img width="1430" alt="Screenshot 2023-01-19 at 2 57 55 PM" src="https://user-images.githubusercontent.com/60198022/213406832-46429f0e-79c1-42ca-97c7-5427a1cc12f2.png">

Builds are failed. Hence, moving this PR to extended testing.

engcom-Delta committed:
Merge remote-tracking branch 'magento/2.4-develop' into indexer-unsubscribe-drop

engcom-Delta commented:
@magento run all tests

engcom-Delta commented:
Hi @fredden,

Thank you for your contribution!

The functional test failures in B2B and EE are different from the last two runs.
Functional tests B2B: run1
<img width="1009" alt="Screenshot 2023-01-24 at 12 23 42 PM" src="https://user-images.githubusercontent.com/51681379/214337184-fdb7998b-31e7-44a1-976e-22c6061814d4.png">
run2
<img width="1023" alt="Screenshot 2023-01-24 at 12 31 14 PM" src="https://user-images.githubusercontent.com/51681379/214337547-480b5faf-ffba-40a8-bee9-2990e95e7805.png">

Functional tests EE: run1
<img width="1023" alt="Screenshot 2023-01-24 at 12 34 05 PM" src="https://user-images.githubusercontent.com/51681379/214338216-67cbf943-bd0f-469f-8b83-daee6197bfee.png">

run2
<img width="1014" alt="Screenshot 2023-01-24 at 12 35 38 PM" src="https://user-images.githubusercontent.com/51681379/214338620-4aa72f16-7f6d-4c1d-8a91-4359903173ea.png">


Raised internal approval request JIRA [ticket](https://jira.corp.adobe.com/browse/AC-7798) for SVC test failure. Waiting for the approvals.

Thank you!

engcom-Delta commented:
Hi @fredden,
We will let you know as soon as we receive jira approval.
Thanks

fredden commented:
@engcom-Delta please can you provide an update on this pull request?

engcom-Charlie commented:
Hi @fredden, we are still waiting for the approvals. Once we got all the required approvals from the internal team, we will proceed further on this PR.

Thank you!

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into indexer-unsubscribe-drop

fredden commented:
@magento run all tests

engcom-Charlie commented:
It seems that bot moved this PR to review hence moving it back to Pending Approval.

fredden commented:
@engcom-Charlie, please can you provide an update on this pull request.

engcom-Echo committed:
Merge branch '2.4-develop' into indexer-unsubscribe-drop

engcom-Echo commented:
@magento run all tests

engcom-Delta commented:
@magento run all tests

engcom-Delta commented:
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`**

**Manual testing scenario:**

1.Login to magento admin
2.Go to System - Index Management
3.Select All indexers from Mass Actions and select Update by Schedule from Actions, click submit.
4.The indexers Status is updated, run php bin/magento indexer:status
5.Go to Catalog - Products and Edit an existing Product Name. Save product
6.Run php bin/magento indexer:status
7.Wait for normal minutely cron to run index or run php  bin/magento cron:run
8.Check the indexer status php bin/magento indexer:status
9.Go to System - Index Management
10.Select All indexers from Mass Actions and select Update on save from Actions, click Submit
11.Check the indexers Status, run php bin/magento indexer:status

**Before:** :heavy_multiplication_x: 

<img width="1431" alt="Screenshot 2024-03-27 at 5 20 46 PM" src="https://github.com/magento/magento2/assets/51681379/29a81040-1381-4d32-ac62-4b2c838b060f">

<img width="1440" alt="Screenshot 2024-03-27 at 5 20 54 PM" src="https://github.com/magento/magento2/assets/51681379/f51a434e-5a01-4e73-8b6f-1bb8d8e0ac00">

<img width="829" alt="Screenshot 2024-03-27 at 5 21 18 PM" src="https://github.com/magento/magento2/assets/51681379/513c23fa-be39-4066-85f6-6e14c5ca95dc">

<img width="833" alt="Screenshot 2024-03-27 at 5 21 28 PM" src="https://github.com/magento/magento2/assets/51681379/b389c6e4-e771-44d5-86f7-4916e2897274">

<img width="1430" alt="Screenshot 2024-03-27 at 5 22 09 PM" src="https://github.com/magento/magento2/assets/51681379/b98d4357-74a5-4ceb-b423-36028d430630">

<img width="1440" alt="Screenshot 2024-03-27 at 5 23 05 PM" src="https://github.com/magento/magento2/assets/51681379/4327a29a-eabf-4857-9234-8544d02694a6">

<img width="1439" alt="Screenshot 2024-03-27 at 5 23 14 PM" src="https://github.com/magento/magento2/assets/51681379/e023c592-bdab-4a1d-8416-01214413ad87">

<img width="823" alt="Screenshot 2024-03-27 at 5 24 37 PM" src="https://github.com/magento/magento2/assets/51681379/4deb8df1-c131-4455-88f1-81ff73d91168">

<img width="1440" alt="Screenshot 2024-03-27 at 5 25 19 PM" src="https://github.com/magento/magento2/assets/51681379/212efecb-3f39-4f48-8e4b-79167d94880d">

<img width="1439" alt="Screenshot 2024-03-27 at 5 25 28 PM" src="https://github.com/magento/magento2/assets/51681379/2a3aaba3-b032-4f54-936f-4d386b016b59">

<img width="778" alt="Screenshot 2024-03-27 at 5 25 44 PM" src="https://github.com/magento/magento2/assets/51681379/c4905385-02f3-427a-b815-c5ba3c208eb0">



**After:** :heavy_check_mark:  

<img width="1435" alt="Screenshot 2024-03-27 at 6 38 28 PM" src="https://github.com/magento/magento2/assets/51681379/fd80ecdd-3a16-4b51-94ca-d3f9835c97f3">

<img width="1440" alt="Screenshot 2024-03-27 at 6 38 44 PM" src="https://github.com/magento/magento2/assets/51681379/e451c671-b6e1-4783-bb24-0adbd3f25fba">

<img width="1434" alt="Screenshot 2024-03-27 at 6 38 53 PM" src="https://github.com/magento/magento2/assets/51681379/436e8dd0-64cf-4c49-b366-aaacd18c0a37">

<img width="836" alt="Screenshot 2024-03-27 at 6 39 12 PM" src="https://github.com/magento/magento2/assets/51681379/ca52feed-d2bd-49e8-9aa5-841793060f45">

<img width="1427" alt="Screenshot 2024-03-27 at 6 40 01 PM" src="https://github.com/magento/magento2/assets/51681379/007b98de-dbad-482a-9bcc-9eac3301a859">

<img width="1439" alt="Screenshot 2024-03-27 at 6 40 11 PM" src="https://github.com/magento/magento2/assets/51681379/e499c63f-8284-44be-995a-089bb5683b42">

<img width="1435" alt="Screenshot 2024-03-27 at 6 41 38 PM" src="https://github.com/magento/magento2/assets/51681379/23b97674-20ae-4f88-afd1-e09dcff8a07d">

<img width="1429" alt="Screenshot 2024-03-27 at 6 41 49 PM" src="https://github.com/magento/magento2/assets/51681379/fa31b9a4-8513-4e05-aa26-2343bd8793d1">

<img width="853" alt="Screenshot 2024-03-27 at 6 42 12 PM" src="https://github.com/magento/magento2/assets/51681379/f9e3cc4b-f38c-4eb5-8975-1aa7581720c0">


Builds are failed. Hence, moving this PR to **Extended Testing.**

engcom-Echo committed:
Merge branch '2.4-develop' into indexer-unsubscribe-drop

engcom-Echo commented:
@magento run all tests

engcom-Echo commented:
@magento run WebAPI Tests,Integration Tests,Functional Tests EE ,Functional Tests CE,Functional Tests B2B

engcom-Echo commented:
Functional Tests CE,Functional Tests EE,Functional Tests B2B failure are different on last two run on same commit.

Functional Tests CE
Run1
<img width="1437" alt="ce-25859" src="https://github.com/magento/magento2/assets/51681435/491abb35-aa1e-4da9-b43f-a7c90c06ef42">

Run2
<img width="1440" alt="Screenshot 2024-03-28 at 1 00 16 PM" src="https://github.com/magento/magento2/assets/51681435/fa214b1c-6366-41ee-bbb8-374282dd68e3">

Functional Tests EE
Run1
<img width="1431" alt="ee-25859" src="https://github.com/magento/magento2/assets/51681435/f9f9f43e-497d-48fe-b5b7-daf80096dc82">

Run2
<img width="1439" alt="Screenshot 2024-03-28 at 1 00 25 PM" src="https://github.com/magento/magento2/assets/51681435/1af6aba4-dcd0-45ad-84c5-ff6c41e60d70">

engcom-Echo commented:
@magento run WebAPI Tests,Integration Tests ,Functional Tests B2B

engcom-Echo commented:
@magento run Functional Tests B2B

engcom-Echo commented:
@magento run Functional Tests B2B
